README.md
> ⚠ Right now the deployed version at NPM only works with node projects. Browser version is comming soon.

[![npm (scoped)](https://img.shields.io/npm/v/@atomic.builders/reactor)](https://www.npmjs.com/package/@atomic-builders/reactor)

# ☢ Reactor

## Reactive-store

Basic _minimalistic REDUX_ made with RxJs

> Production ready **observable state management**

## 🎯 Motivation

✅ Avoid the use of heavy and complex libraries like _NgRX_

✅ Less than 100 lines [store.ts](https://github.com/AtomicBuilders/reactor/blob/master/src/store.ts)

## Installation

```bash
npm i @atomic.builders/reactor
```

## ✨ Sample use

```typescript
const initialBasket: Basket = { client: '', items: [], status: '' };
const basket$ = new Store<Basket>(initialBasket);

// BASIC USAGE
// get snapshot
const currentBasket: Basket = basket$.getState();
console.log(currentBasket);
// observe changes
basket$.getState$().subscribe({
  next: basket => console.log({ basket }),
});
// observe selected changes
basket$
  .select$(basket => basket.status)
  .subscribe({
    next: status => console.log({ status }),
  });
// setting direct state
basket$.setState({ client: '', items: [], status: 'EMPTY' });

// WHITH ACTIONS
// observe actions
basket$.getActions$().subscribe({
  next: action => console.log({ action }),
});
// dispatch simple action
const setClientAction: Action<Basket> = new Action<Basket>('SET_CLIENT', {
  client: 'John Doe',
  status: 'EMPTY',
});
basket$.dispatch(setClientAction);

// WITH A REDUCER
// dispatch action with payload and reducer
const itemPayload: Item = {
  name: 'An ACME thing',
  units: 19,
  unitPrice: 71,
};
const addItemReducer = (basket: Basket, payload: unknown) => {
  basket.items = [...basket.items, payload as Item];
  basket.status = 'FILLED';
  return basket;
};
const addItemAction: Action<Basket> = new Action<Basket>('ADD_ITEM', itemPayload, addItemReducer);
basket$.dispatch(addItemAction);
```

## ⚙ Workflows

### 👨‍💻 Dev Workflow

While developing, make sure to install the recommended extensions for a better dev experience.

Run `npm run test:watch` it will run test after each change. Ideal for TDD or testing just in time.

Run `npm run test` it will run all test once and stops. Default for CI/CD most common environments.

If you want also the coverage report then use `npm run test:coverage` .

## 🛠 Tools

### 📦 Commits and release

- Use [standard-version](https://www.npmjs.com/package/standard-version) to produce a changelog file from [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)

### 💅 Code style with Prettier

- Installed and configured prettier

> Recommended [prettier extension](https://github.com/prettier/prettier-vscode)

### 📐 Code linting with esLint

- Installed and configured eslint to work with prettier

> Recommended [esLint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

### 🧪 Code tested with Jest

- Installed and configured **jest** to run specs
- Configured to conform with **eslint**
- Uses `ts-jest` to work natively with **TypeScript**

## 👨 Created by Alberto Basalo

> [@albertobasalo](https://twitter.com/albertobasalo)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.2](https://github.com/AtomicBuilders/reactor/compare/v2.0.1...v2.0.2) (2021-04-09)

### [2.0.1](https://github.com/AtomicBuilders/reactor/compare/v2.0.0...v2.0.1) (2021-04-09)

## [2.0.0](https://github.com/AtomicBuilders/reactor/compare/v1.2.7...v2.0.0) (2021-04-06)


### ⚠ BREAKING CHANGES

* allow partial changes, generate acation from setstate

### Features

* allow partial changes, generate acation from setstate ([7c4579e](https://github.com/AtomicBuilders/reactor/commit/7c4579e9924acf400209a1e503bd682bfc5f4b75))

### [1.2.7](https://github.com/AtomicBuilders/reactor/compare/v1.2.6...v1.2.7) (2021-04-06)

### [1.2.6](https://github.com/AtomicBuilders/reactor/compare/v1.2.5...v1.2.6) (2021-04-05)

### [1.2.5](https://github.com/AtomicBuilders/reactor/compare/v1.2.4...v1.2.5) (2021-04-05)

### [1.2.4](https://github.com/AtomicBuilders/reactor/compare/v1.2.3...v1.2.4) (2021-04-05)

### [1.2.3](https://github.com/AtomicBuilders/reactor/compare/v1.2.2...v1.2.3) (2021-04-05)

### [1.2.2](https://github.com/AtomicBuilders/reactor/compare/v1.2.1...v1.2.2) (2021-04-05)

### [1.2.1](https://github.com/AtomicBuilders/reactor/compare/v1.2.0...v1.2.1) (2021-04-05)

## [1.2.0](https://github.com/AtomicBuilders/reactor/compare/v1.1.1...v1.2.0) (2021-04-05)


### Features

* rename to reactor ([47924b7](https://github.com/AtomicBuilders/reactor/commit/47924b7dcdf730e139f5150322ea0da686258ed9))

### [1.1.1](https://github.com/AtomicBuilders/rxjs-store/compare/v1.1.0...v1.1.1) (2021-04-04)

## 1.1.0 (2021-04-04)

### Features

- action reducers ([fc6b151](https://github.com/AtomicBuilders/rxjs-store/commit/fc6b151838dfbb4268e9f95e1952da8d31316181))
- add actions ([022e754](https://github.com/AtomicBuilders/rxjs-store/commit/022e7549c0327bbdcca91392b5f8ec22a00221a7))
- can get and emit changes ([582edd5](https://github.com/AtomicBuilders/rxjs-store/commit/582edd56817347151952bdde797769c56ae465d2))
- immutable observable ([1aa0a25](https://github.com/AtomicBuilders/rxjs-store/commit/1aa0a2516d4f351ad11b7f67f0532cfe86359e9e))
- private inmutable state ([b21163a](https://github.com/AtomicBuilders/rxjs-store/commit/b21163a582de7e5c12b1f60bb430ed9fdc840307))
- selector ([3c8c192](https://github.com/AtomicBuilders/rxjs-store/commit/3c8c1924e168d0d0e4d46710815bd50f72baca30))
- store observable ([1b6f103](https://github.com/AtomicBuilders/rxjs-store/commit/1b6f1034c381e32010e9b1dbbe5b408cad762ddc))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.


samples/basic.ts
import { Basket } from '../samples/basket';
import { Item } from '../samples/item';
import { Action } from '../src/models/action';
import { Store } from '../src/store';

const initialBasket: Basket = { client: '', items: [], status: '' };
const basket$ = new Store<Basket>(initialBasket);

// BASIC USAGE
// get snapshot
const currentBasket: Basket = basket$.getState();
console.log(currentBasket);
// observe changes
basket$.getState$().subscribe({
  next: basket => console.log({ basket }),
});
// observe selected changes
basket$
  .select$(basket => basket.status)
  .subscribe({
    next: status => console.log({ status }),
  });
// setting direct state
basket$.setState({ client: '', items: [], status: 'EMPTY' });

// WHITH ACTIONS
// observe actions
basket$.getActions$().subscribe({
  next: action => console.log({ action }),
});
// dispatch simple action
const setClientAction: Action<Basket> = new Action<Basket>('SET_CLIENT', {
  client: 'John Doe',
  status: 'EMPTY',
});
basket$.dispatch(setClientAction);

// WITH A REDUCER
// dispatch action with payload and reducer
const itemPayload: Item = {
  name: 'An ACME thing',
  units: 19,
  unitPrice: 71,
};
const addItemReducer = (basket: Basket, payload: unknown) => {
  basket.items = [...basket.items, payload as Item];
  basket.status = 'FILLED';
  return basket;
};
const addItemAction: Action<Basket> = new Action<Basket>('ADD_ITEM', itemPayload, addItemReducer);
basket$.dispatch(addItemAction);
